### PR TITLE
Added option restrict-upload-dir

### PR DESCRIPTION
**Pull request by jonasdiemer**
To fix https://github.com/svenstaro/miniserve/issues/842

### CODE REVIEW

**Comment by svenstaro** [comment]
Cool, before I do a proper review, could you add a test to check that other directories are now indeed forbidden?

**Comment by jonasdiemer** [comment]
Sure, let me have a look into testing in Rust :)

**Comment by svenstaro** [comment]
You can take a look at the other tests concerning the upload. It should be somewhat straight forward. Just remember: If you messed up and the test doesn't pick it up, this is a security issue so we need to be really careful here. :)

**Comment by jonasdiemer** [comment]
@svenstaro thanks for the suggestion. Fixed an issue (running tests is actually helpful :)) and added a test similar to existing one `uploading_files_is_prevented`.

**Review by svenstaro — CHANGES_REQUESTED** [review]
Good stuff! This will probably require a few rounds of reviews until we get there but i think for your first foray into Rust this is already looking pretty good.

File: tests/upload_files.rs [hunk 1]
```diff
@@ -80,6 +80,44 @@ fn uploading_files_is_prevented(server: TestServer) -> Result<(), Error> {
...
+    let client = Client::new();
+    // Ensure uploading fails and returns an error
+    assert!(client
+        .post(server.url().join("/upload?path=/")?)
+        .multipart(form)
+        .send()?
+        .error_for_status()
+        .is_err());
+
+    // After uploading, check whether the uploaded file is now getting listed.
```
[the remark this answers is not in the document]
Comment (jonasdiemer) [reply tests/upload_files.rs]: Copy&paste error at the place where I copied from, fixed.

File: tests/upload_files.rs [hunk 2]
```diff
@@ -80,6 +80,44 @@ fn uploading_files_is_prevented(server: TestServer) -> Result<(), Error> {
...
+    let body = reqwest::blocking::get(server.url())?;
+    let parsed = Document::from_read(body)?;
+    assert!(!parsed.find(Text).any(|x| x.text() == test_file_name));
+
```
[the remark this answers is not in the document]
Comment (jonasdiemer) [reply tests/upload_files.rs:111]: Added another test.

File: src/file_upload.rs [hunk 3]
```diff
@@ -171,6 +171,19 @@ pub async fn upload_file(
         ContextualError::IoError("Failed to resolve path served by miniserve".to_string(), e)
     })?;
 
+
+    // Disallow paths outside of restricted directories
+    // TODO: Probably not the most rust-ic style...
+    if !conf.restrict_upload_dir.is_empty() {
+        let upl_path = upload_path.clone().into_os_string().into_string().unwrap();
```
[the remark this answers is not in the document]
Comment (jonasdiemer) [reply src/file_upload.rs:178]: Fully agree to how it reads.

One option I am seeing is to use PathBuf for the --restrict-upload-dir argument instead of String and then handle it as a path. @svenstaro, would this make sense?

File: src/args.rs [hunk 4]
```diff
@@ -111,6 +111,10 @@ pub struct CliArgs {
     #[clap(short = 'u', long = "upload-files")]
     pub file_upload: bool,
 
+    /// Restrict upload directories
+    #[clap(long = "restrict-upload-dir")]
```
[the remark this answers is not in the document]
Comment (jonasdiemer) [reply src/args.rs:115]: Thanks, makes total sense, done.

**Comment by jonasdiemer** [comment]
Reworked the solution to use PathBuf and with this also fixed subdirectories not working.

@svenstaro I'd appreciate some feedback (esp. on improving the style). Thanks again! :)

File: src/file_upload.rs [hunk 5]
```diff
@@ -171,6 +171,18 @@ pub async fn upload_file(
         ContextualError::IoError("Failed to resolve path served by miniserve".to_string(), e)
     })?;
 
+
+    // Disallow paths outside of restricted directories
+    // TODO: Probably not the most rust-ic style...
+    let upload_allowed = conf.restrict_upload_dir.is_empty() || 
+        conf.restrict_upload_dir.iter().any(|s| upload_path.starts_with(s)); 
+
+    if !(upload_allowed) {
```
Comment (svenstaro) [inline src/file_upload.rs:180]:
```suggestion
    if !upload_allowed {
```

File: src/file_upload.rs [hunk 6]
```diff
@@ -171,6 +171,18 @@ pub async fn upload_file(
         ContextualError::IoError("Failed to resolve path served by miniserve".to_string(), e)
     })?;
 
+
+    // Disallow paths outside of restricted directories
+    // TODO: Probably not the most rust-ic style...
+    let upload_allowed = conf.restrict_upload_dir.is_empty() ||
```
Comment (svenstaro) [inline src/file_upload.rs:177]:
```suggestion
    // Disallow paths outside of restricted directories
    let upload_allowed = conf.restrict_upload_dir.is_empty() || 
```

File: src/args.rs [hunk 7]
```diff
@@ -111,6 +111,10 @@ pub struct CliArgs {
     #[clap(short = 'u', long = "upload-files")]
     pub file_upload: bool,
 
+    /// Restrict upload directories
+    #[clap(long = "restrict-upload-dir", requires = "file-upload", value_hint = ValueHint::FilePath)]
+    pub restrict_upload_dir: Vec<PathBuf>,
```
Comment (svenstaro) [inline src/args.rs:116]:
```suggestion
    /// Allowed upload directories
    ///
    /// If this is set, uploads are only allowed into the provided directories. 
    #[clap(long, requires = "file-upload", value_hint = ValueHint::FilePath)]
    pub allowed_upload_dir: Vec<PathBuf>,
```

Comment (svenstaro) [reply src/args.rs:116]: I believe this explanation and naming makes more sense. Please change it like that throughout your code unless you have a better suggestion for the naming.

**Review by svenstaro — CHANGES_REQUESTED** [review]

File: tests/upload_files.rs [hunk 8]
```diff
@@ -72,14 +72,96 @@ fn uploading_files_is_prevented(server: TestServer) -> Result<(), Error> {
...
+    // After uploading, check whether the uploaded file is NOT getting listed.
     let body = reqwest::blocking::get(server.url())?;
     let parsed = Document::from_read(body)?;
     assert!(!parsed.find(Text).any(|x| x.text() == test_file_name));
 
     Ok(())
 }
 
+// This test runs the server with --restrict-upload-dir argument and 
+// checks that file upload to a different directory is actually prevented.
```
Comment (svenstaro) [inline tests/upload_files.rs:84]: Prefer docstrings `///` on top of functions instead of normal comments `//`.

File: tests/upload_files.rs (hunk 2 above)
[the remark this answers is not in the document]
Comment (svenstaro) [reply tests/upload_files.rs:113]: I think perhaps we could merge the tests using another rstest parameter (success/failure)? We do this in some other places in the tests as well. It always depends on much the code would diverge. Can you check how that would look and whether it makes sense in this case?

File: src/args.rs (hunk 7 above)
Comment (jonasdiemer) [reply src/args.rs:116]: Yes, this is clearer.

File: tests/upload_files.rs (hunk 2 above)
[the remark this answers is not in the document]
Comment (jonasdiemer) [reply tests/upload_files.rs:113]: There is not that much duplication in the two tests - I have removed some redundant parts in the `uploading_files_is_restricted` (eg. checking if file exists). So in my view merging would make the tests significantly less easy to follow.

**Comment by svenstaro** [comment]
I'd be happy to merge as-is. Please update the `README.md` with the new output of `miniserve --help` and fix the formatting issue and we're good to merge.

File: src/config.rs [hunk 9]
```diff
@@ -87,6 +87,9 @@ pub struct MiniserveConfig {
     /// Enable file upload
     pub file_upload: bool,
 
+    /// Restrict file upload dirs
```
Comment (svenstaro) [inline src/config.rs:90]:
```suggestion
    /// List of allowed upload directories
```

File: src/file_upload.rs [hunk 10]
```diff
@@ -171,6 +171,17 @@ pub async fn upload_file(
         ContextualError::IoError("Failed to resolve path served by miniserve".to_string(), e)
     })?;
 
+
+    // Disallow paths outside of restricted directories
+    let upload_allowed = conf.allowed_upload_dir.is_empty() || 
+        conf.allowed_upload_dir.iter().any(|s| upload_path.starts_with(s));
```
Comment (svenstaro) [inline src/file_upload.rs:177]:
```suggestion
    // Disallow paths outside of allowed directories
    let upload_allowed = conf.allowed_upload_dir.is_empty() || 
        conf.allowed_upload_dir.iter().any(|s| upload_path.starts_with(s)); 
```

File: src/args.rs [hunk 11]
```diff
@@ -111,6 +111,12 @@ pub struct CliArgs {
     #[clap(short = 'u', long = "upload-files")]
     pub file_upload: bool,
 
+    /// Allowed upload directories (together with -u)
+    /// 
+    /// If this is set, uploads are only allowed into the provided directories. 
+    #[clap(long, requires = "file-upload", value_hint = ValueHint::FilePath)]
```
Comment (svenstaro) [inline src/args.rs:117]: If you like, you can choose a short letter for this as well as it'd be annoying to write it out many times otherwise.

File: tests/upload_files.rs [hunk 12]
```diff
@@ -72,14 +72,96 @@ fn uploading_files_is_prevented(server: TestServer) -> Result<(), Error> {
...
+    let parsed = Document::from_read(body)?;
+    assert!(!parsed.find(Text).any(|x| x.text() == test_file_name));
+
+    Ok(())
+}
+
+/// This tests that we can upload files to the directory specified by --allow-upload-dir
+#[rstest]
+#[case(server(&["-u",  "--allowed-upload-dir", "someDir"]), "someDir")]
+#[case(server(&["-u",  "--allowed-upload-dir", "someDir/some_sub_dir"]), "someDir/some_sub_dir")]
```
Comment (svenstaro) [inline tests/upload_files.rs:120]: Actually could you add a test case that shows that it also works when multiple `--allowed-upload-dir` are provided?

**Review by svenstaro — CHANGES_REQUESTED** [review]
Found a few tiny things but once we have that, we're good to go. Great job so far!

File: tests/upload_files.rs [hunk 13]
```diff
@@ -72,14 +72,96 @@ fn uploading_files_is_prevented(server: TestServer) -> Result<(), Error> {
...
+fn uploading_files_to_allowed_dir_works(
+    #[case] server: TestServer,
+    #[case] upload_dir: &str,
+) -> Result<(), Error> {
+    let test_file_name = "uploaded test file.txt";
+
+    // Create test directory
+    use std::fs::create_dir_all;
```
Comment (svenstaro) [inline tests/upload_files.rs:128]: We already have this in another place in this file. Just hoist this import up to global scope and remove it in the other place as well.

File: src/args.rs (hunk 11 above)
Comment (jonasdiemer) [reply src/args.rs:117]: I'd go for 'A' (as in allowed...), but I am wondering if we might just want to re-purpose `-u`...

Comment (svenstaro) [reply src/args.rs:117]: You mean something like `-u` alone would allow for uploading everywhere and then `-u /tmp/lol` would allow uploading only to that dir? If that's what you're thinking, I appreciate the conciseness but I can't help but wonder whether that wouldn't make `-u` pretty hard to parse (is it even still possible?) and somewhat confusing to users perhaps. Thoughts?

File: tests/upload_files.rs [hunk 14]
```diff
@@ -72,14 +72,96 @@ fn uploading_files_is_prevented(server: TestServer) -> Result<(), Error> {
...
+    let form = form.part("file_to_upload", part);
+
+    let client = Client::new();
+    // Ensure uploading fails and returns an error
+    assert!(client
+        .post(server.url().join("/upload?path=/")?)
+        .multipart(form)
+        .send()?
+        .error_for_status()
+        .is_err());
```
Comment (svenstaro) [inline tests/upload_files.rs:107]: I was just thinking, perhaps check that this is actually forbidden. Otherwise, this test might fail for a 404 and we wouldn't catch it. Maybe also check for the precise error message?

File: src/args.rs (hunk 11 above)
Comment (svenstaro) [reply src/args.rs:117]: `-A` is a fine choice, I think.

**Review by svenstaro — CHANGES_REQUESTED** [review]
Hey, I think we're almost there now. Only some tiny stuff now. I'd like to put this into the next miniserve release, too! Do you have some time to finish it up soon so we can land it with the release?

File: src/args.rs (hunk 11 above)
Comment (jonasdiemer) [reply src/args.rs:117]: > You mean something like `-u` alone would allow for uploading everywhere and then `-u /tmp/lol` would allow uploading only to that dir? If that's what you're thinking, I appreciate the conciseness but I can't help but wonder whether that wouldn't make `-u` pretty hard to parse (is it even still possible?) and somewhat confusing to users perhaps. Thoughts?

Exactly, that was my thinking (after seeing so many command line shorts already taken ;)). Not sure about the parsing, though, I'd give it a try...

Comment (jonasdiemer) [reply src/args.rs:117]: Played around a bit, it looks like it would be relatively easy to parse by using `min_values = 0`:

```rust
    #[clap(short = 'u', long,
     value_hint = ValueHint::FilePath, 
     min_values = 0)]
    pub upload-files: Option<Vec<PathBuf>>,
```

This would yield `None` in case the argument is missing, `Some([])` in case the argument is used without parameters and `Some(['some_dir'])` in case a directory is supplied. So doesn't look to complicated in my view.

@svenstaro what do you think? I would remodel the code to have `-u, --upload-files [<ALLOWED_UPLOAD_DIR>...]` and save the extra argument `-A/--allowed_upload-dir`

Comment (svenstaro) [reply src/args.rs:117]: Does it parse this correctly? `-u --some-other-flag`? Would it try to parse the stuff behind `-u` as a `PathBuf`?

Comment (jonasdiemer) [reply src/args.rs:117]: Just tried `-u --verbose` and it worked as expected (i.e. verbose output and `allowed_upload_dir == Some([])` (i.e. it would allow uploads everywhere)

File: tests/upload_files.rs (hunk 12 above)
Comment (jonasdiemer) [reply tests/upload_files.rs]: Done.

File: tests/upload_files.rs (hunk 14 above)
Comment (jonasdiemer) [reply tests/upload_files.rs]: Yes, that makes more sense.

File: src/args.rs (hunk 11 above)
Comment (jonasdiemer) [reply src/args.rs]: I've just committed a version that uses `-u` to define allowed directories. @svenstaro, please check it out and let me know what you think (I am also fine to use the previous version with `-A` as a shortcut).

**Comment by jonasdiemer** [comment]
> Hey, I think we're almost there now. Only some tiny stuff now. I'd like to put this into the next miniserve release, too! Do you have some time to finish it up soon so we can land it with the release?

Except for the decision for using -u or -A (and some cleanup in case we go with -u), it should be ready.

File: src/args.rs (hunk 11 above)
Comment (Atreyagaurav) [reply src/args.rs]: I like the `-u` with `min_values`. Makes more sense. Might want to try how it does in windows shell. It should work in bash because we've to use `./-path` when a file starts with `-` (e.g. `touch ./-h`) otherwise it'll be parsed as cli argument.

Comment (jonasdiemer) [reply src/args.rs]: Unfortunately, I cannot try on Windows.

Directories starting with `-` would also need the `./` prefix, which doesn't work with my implementation (yet) - will have a look.

**Comment by svenstaro** [comment]
Please run `cargo fmt` to fix CI.

**Comment by jonasdiemer** [comment]
Fixed formatting and also sanitized paths so that specification of `-u ./some_dir' also works.

**Comment by svenstaro** [comment]
There's a small lint to fix there.

**Comment by jonasdiemer** [comment]
Looks like CICD on Windows fails, might have to do with the tests that include a dash in the directory name. Unfortunately, I don't have a windows machine to test on...

**Comment by Atreyagaurav** [comment]
I also don't have a Windows machine to build miniserve to test, but I made a colleague try to make a directory named "-h" and it was a success (I tried both with "New Folder" button and `mkdir` in cmd. `cd -h` & `cd ./-h`worked fine, but `dir ./-h` gave `Invalid switch - "h".`

I don't know if it's a good idea, but we might be able to use `cfg` to only test that on Linux? At least for now...

**Comment by svenstaro** [comment]
I'd be fine if we don't run that test on Windows.

**Comment by jonasdiemer** [comment]
>

File: tests/upload_files.rs [hunk 15]
```diff
@@ -117,23 +117,27 @@ fn uploading_files_is_restricted(#[case] server: TestServer) -> Result<(), Error
...
-#[case(server(&["-u", "someDir/some_sub_dir"]), vec!["someDir/some_sub_dir"])]
+#[case(server(&["-u", "someDir"]), vec!["someDir".to_string()])]
+#[case(server(&["-u", "./-someDir"]), vec!["./-someDir".to_string()])]
+#[case(server(&["-u", "someDir/some_sub_dir"]), vec!["someDir/some_sub_dir".to_string()])]
 #[case(server(&["-u", "someDir/some_sub_dir", "-u", "someDir/some_other_dir"]), 
-       vec!["someDir/some_sub_dir", "someDir/some_other_dir"])]
+       vec!["someDir/some_sub_dir".to_string(), "someDir/some_other_dir".to_string()])]
 fn uploading_files_to_allowed_dir_works(
     #[case] server: TestServer,
-    #[case] upload_dirs: Vec<&str>,
```
Comment (svenstaro) [inline tests/upload_files.rs]: Does it work if you take these as `Path` or `OsPath`?

File: tests/upload_files.rs [hunk 16]
```diff
@@ -72,14 +73,103 @@ fn uploading_files_is_prevented(server: TestServer) -> Result<(), Error> {
...
+
+    let client = Client::new();
+    // Ensure uploading fails and returns an error
+    assert_eq!(
+        500,
+        client
+            .post(server.url().join("/upload?path=/")?)
+            .multipart(form)
+            .send()?
+            .status()
```
Comment (svenstaro) [inline tests/upload_files.rs:107]: I just found this. We should never return 500s. This is a clear case of FORBIDDEN.

Comment (jonasdiemer) [reply tests/upload_files.rs:107]: Yes, fully agree. However, that's a behavior that I've inherited from how miniserve treats other errors (e.g. uploading a file again without `overwrite_files` option), so I would think this should be fixed in a separate PR...

Comment (svenstaro) [reply tests/upload_files.rs:107]: That's a fair argument to make. Would you be interested in making such a follow-up?

**Comment by svenstaro** [comment]
It seems like there's a conflict with the recent QR stuff merge. Could you rebase?

**Comment by jonasdiemer** [comment]
Rebase done, still need to fix the failing test for windows platform.

File: tests/upload_files.rs [hunk 17]
```diff
@@ -72,14 +73,101 @@ fn uploading_files_is_prevented(server: TestServer) -> Result<(), Error> {
...
+#[case(server(&["-u", "someDir"]), vec!["someDir".to_string()])]
+#[case(server(&["-u", "./-someDir"]), vec!["./-someDir".to_string()])]
+#[case(server(&["-u", if cfg!(windows) {r"someDir\some_sub_dir"} else {"someDir/some_sub_dir"}]),
+  vec!["someDir/some_sub_dir".to_string()])]
+#[case(server(&["-u", if cfg!(windows) {r"someDir\some_sub_dir"} else {"someDir/some_sub_dir"}, 
+                "-u", if cfg!(windows) {r"someDir\some_other_dir"} else {"someDir/some_other_dir"}]), 
+       vec!["someDir/some_sub_dir".to_string(), "someDir/some_other_dir".to_string()])]
+fn uploading_files_to_allowed_dir_works(
+    #[case] server: TestServer,
+    #[case] upload_dirs: Vec<String>,
```
Comment (svenstaro) [inline tests/upload_files.rs:129]: I still don't really understand why you can't use a `std::path::Path` here as it will take care of the conversion for you on the operating systems. Wouldn't that be more convenient?

Comment (svenstaro) [reply tests/upload_files.rs]: We should probably change other tests as well. :)

Comment (svenstaro) [reply tests/upload_files.rs]: Since it still doesn't work, I'd also be ok with you simply disabling the failing test cases on Windows.

Comment (jonasdiemer) [reply tests/upload_files.rs]: Yup, and kind of hard to make progress without access to a windows machine. Not sure I'll find time to setup something (e.g. on Azure/CodeSpaces)... 

@svenstaro, I don't know how I would disable a test-case for a specific platform. Could you give me a pointer?

Comment (jonasdiemer) [reply tests/upload_files.rs]: Also noticed, that the feature doesn't work on windows with subdirectories (likely due to `\` vs `/`) - found out by running miniserve.exe from publish_release action on an old windows laptop. Pushed a fix, waiting for Action workflow to verify.

Comment (jonasdiemer) [reply tests/upload_files.rs]: Looks like I found a solution - latest commit passes CI on all platforms.

I assume the warning (the following packages contain code that will be rejected by a future version of Rust: winapi v0.2.8) has nothing to do with my PR.

From my perspective, the PR should be ready to merge now - please let me know if I overlooked something.

Re the 500 errors, I might have a stab at that if time permits.

Comment (jonasdiemer) [reply tests/upload_files.rs]: Well, the error handling was easier to fix than I thought - fixed it (also for the case when `overwrite_files` is not set, i.e. `DuplicateFileError`).

**Comment by svenstaro** [comment]
At last! Outstanding work getting it to this state. It took a while but I hope you appreciate that for security stuff we need to go the extra mile. :)

**Comment by jonasdiemer** [comment]
Thanks a lot for your guidance (and patience), I leaned a lot and it was fun!
